test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl

```diff
@@ -0,0 +1,170 @@
+
+// channels for doing efficient console I/O
+var stdinBin  = openfd(0).reader(iokind.native, locking=false,
+                           hints=ioHintSet.fromFlag(QIO_CH_ALWAYS_UNBUFFERED)),
+    stdoutBin = openfd(1).writer(iokind.native, locking=false,
+                           hints=ioHintSet.fromFlag(QIO_CH_ALWAYS_UNBUFFERED));
+
+proc main(args: [] string) {
+  // set up the 'pairCmpl' map
+  const chars = eol..<maxChars;
+  forall (i,j) in {chars, chars} do
+    pairCmpl[join(i,j)] = join(cmpl(j), cmpl(i));
+
+  // variables for reading into a dynamically growing buffer
+  var buffCap = readSize,       // capacity of our input buffer
+      buffDom = {0..<buffCap},  // index set for the input buffer
+      buff: [buffDom] uint(8),  // the buffer itself
+      readPos = 0;              // the current read position
+
+  do {
+    // read 'readSize' new characters
+    var newChars = stdinBin.readBinary(c_ptrTo(buff[readPos]), readSize),
+        nextSeqStart: int;
+
+    // if the new characters contain the start of the next sequence,
+    while findSeqStart(buff, readPos, newChars, nextSeqStart) {
+      // process this one
+      revcomp(buff, nextSeqStart);
+
+      // then shift the next sequence to the start of the buffer
+      newChars -= nextSeqStart - readPos + 1;
+
+      serial (nextSeqStart < newChars) do
+        forall j in 0..newChars do
+          buff[j] = buff[j+nextSeqStart];
+
+      // and reset to see whether there's another sequence ahead
+      readPos = 1;
+    }
+
+    // update the position to read to next
+    readPos += newChars;
+
+    // if we're about to run out of space, grow the buffer by 2x
+    if readPos + readSize > buffCap {
+      buffCap *= 2;
+      buffDom = {0..<buffCap};
+    }
+  } while newChars;
+
+  // if anything remains, process it
+  if readPos then revcomp(buff, readPos);
+}
+
+proc revcomp(seq, size) {
+  param chunkSize = linesPerChunk*cols;  // the size of the chunks to deal out
+
+  // compute how big the header is
+  var headerSize = 1;
+  while seq[headerSize - 1] != eol {
+    headerSize += 1;
+  }
+
+  // write out the header
+  stdoutBin.writeBinary(c_ptrTo(seq[0]), headerSize);
+
+  // set up the atomic variables we'll use to coordinate between tasks
+  var charsLeft, charsWritten: atomic int = size - (headerSize + 1);
+
+  // create a task per core
+  coforall tid in 0..<here.maxTaskPar {
+    var myBuff: [0..<chunkSize] uint(8);  // the task's buffer
+
+    var myStartChar = charsLeft.fetchSub(chunkSize);
+    while myStartChar >= 0 {
+      const myChunkSize = min(chunkSize, myStartChar + 1),
+            lastLineChars = myStartChar % cols,
+            lastLineGaps = cols - (lastLineChars + 1);
+
+      var cursor = myStartChar + headerSize,
+          chunkLeft = myChunkSize,
+          chunkPos = 0;
+
+      if lastLineGaps == 0 {
+        revcomp(chunkPos, cursor, chunkLeft, myBuff, seq);
+      } else {
+        while chunkLeft >= cols {
```

Comment:
I think this else-branch might benefit from a brief comment explaining its purpose. Or a comment somewhere briefly describing the "last line" issue.

Reply:
You may be incorrectly assuming that I actually understand this code... :D  I've put in at least a very high-level description, though.